Cauldron/Assets/Scripts/ExperimentController.cs

```diff
@@ -0,0 +1,277 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using CauldronCodebase;
+using UnityEngine;
+using UnityEngine.UI;
+using Random = UnityEngine.Random;
+
+public class ExperimentController : MonoBehaviour
+{
+    [SerializeField] private RecipeBook recipeBook;
+    [SerializeField] private AnimalsFilter animalsFilter;
+    [SerializeField] private RootsFilter rootsFilter;
+    [SerializeField] private MushroomsFilter mushroomsFilter;
+    [SerializeField] private PlantsFilter plantsFilter;
+    [SerializeField] private Button nextPage;
+    [SerializeField] private Button prevPage;
+
+    public List<Ingredients[]> currentRecipes = new List<Ingredients[]>();
+    public List<AttemptEntry> attemptEntries;
+    public List<WrongPotion> wrongPotions;
+
+    private const int MaxFilterSelection = 2;
+    private const int ZeroFilter = 0;
+    private const int FirstFilter = 1;
+    private const int SecondFilter = 2;
+    
+    private IngredientsData.Ingredient maiIngredient;
+    private IngredientsData.Ingredient spareIngredient;
+    private bool isFindWrongRecipe;
+    private int countFilter;
+    private int currentPage;
+
+    public int MaxTotalPages => 5;
+
+    private void OnEnable()
+    {
+        nextPage.onClick.AddListener(OnNextPage);
+        prevPage.onClick.AddListener(OnPrevPage);
+        
+        animalsFilter.AddedFilter += OnUpdateButtonFilter;
+        rootsFilter.AddedFilter += OnUpdateButtonFilter;
+        mushroomsFilter.AddedFilter += OnUpdateButtonFilter;
+        plantsFilter.AddedFilter += OnUpdateButtonFilter;
+        animalsFilter.Show += OnUpdateFilter;
+        rootsFilter.Show += OnUpdateFilter;
+        mushroomsFilter.Show += OnUpdateFilter;
+        plantsFilter.Show += OnUpdateFilter;
+    }
+
+    private void OnDisable()
+    {
+        nextPage.onClick.RemoveListener(OnNextPage);
+        prevPage.onClick.RemoveListener(OnPrevPage);
+        
+        animalsFilter.AddedFilter -= OnUpdateButtonFilter;
+        rootsFilter.AddedFilter -= OnUpdateButtonFilter;
+        mushroomsFilter.AddedFilter -= OnUpdateButtonFilter;
+        plantsFilter.AddedFilter -= OnUpdateButtonFilter;
+        animalsFilter.Show -= OnUpdateFilter;
+        rootsFilter.Show -= OnUpdateFilter;
+        mushroomsFilter.Show -= OnUpdateFilter;
+        plantsFilter.Show -= OnUpdateFilter;
+    }
+
+    public void ResetPages()
+    {
+        currentPage = 0;
+    }
+
+    public void RecordAttempt(WrongPotion mix)
+    {
+        if (wrongPotions is null)
+        {
+            wrongPotions = new List<WrongPotion>(10);
+        }
+
+        if (!wrongPotions.Any(wrongRecipe => wrongRecipe.IngredientsList.All(mix.IngredientsList.Contains)))
```

Comment:
This is where we also check ingredients in one line, but it's not really optimal (basically it does what you did in Recipe class, only it looks shorter and has more allocations (this is LINQ))